### PR TITLE
fix: paymentMethodType in DynamicFields

### LINKED ISSUE #1644: fix paymentMethodType in DynamicFields

### PR DESCRIPTION
**Pull request by Shivam25092001**
## Type of Change

- [X] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

#### Bug:
- Bank debit/transfer types come through with a suffix (`ach_debit`, `sepa_transfer`), but Superposition and the `sdk-configs` payment_methods API key on the canonical name (`ach`, `sepa`). Passing the raw suffixed value into `getEligibleConnectorsFromPaymentMethods` and `buildSuperpositionBaseContext` made those lookups miss, yielding empty/incorrect required fields.

#### Fix:
- Fixed `payment_method_type` in `DynamicFields.
- used `PaymentUtils.getPaymentMethodName` for accessing the `payment_method_type`'s name.
- affects some bank debit and bank transfer types. Cards/wallets remains unaffected.

## How did you test it?

| Before                                                                                                                              | After                                                                                                                               |
|-------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------|
| `multibanco_transfer`                                                                                                               | `multibanco`                                                                                                                        |
| image  | image  |
| image | image |

## Checklist

- [X] I ran `npm run re:build`
- [X] I reviewed submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Comment by XyneSpaces** [comment]
[should-fix] Suffix stripping logic may not cover all cases

The `configPaymentMethodType` normalization at lines 179-203 uses `getPaymentMethodName` to strip suffixes like `_debit` and `_transfer`. Ensure this covers:
1. All payment method variations currently in use
2. Future payment methods that may follow different naming conventions

Consider adding a fallback mechanism or logging when the normalized name doesn't match any entries in the Superposition config, to aid debugging lookup misses.

**Comment by XyneSpaces** [comment]
## Review Summary

### [should-fix] Potential edge case in payment method name normalization

**File:** `src/Components/DynamicFields.res` (lines 179-185)

The change introduces `PaymentUtils.getPaymentMethodName` to strip suffixes from payment method types. While this fixes the immediate issue with `ach_debit` → `ach` mapping, there are potential concerns:

1. **Silent fallback behavior**: If `paymentMethodType` is `"bank_debit"` or `"bank_transfer"` but the `paymentMethodName` doesn't contain the expected suffix, the function returns the name unchanged. This could lead to silent failures where the lookup still misses.

2. **String-based matching fragility**: The solution relies on string suffix matching rather than a type-safe mapping. Future payment method types with similar naming patterns could inadvertently trigger incorrect transformations.

**Suggested improvement:**
Consider adding explicit validation/logging when the expected suffix is not found:

```rescript
let getPaymentMethodName = (~paymentMethodType, ~paymentMethodName, ~logger) => {
  if paymentMethodType == "bank_debit" {
    let normalized = paymentMethodName->String.replace("_debit", "")
    if normalized == paymentMethodName {
      logger.warning("Expected _debit suffix not found in: " ++ paymentMethodName)
    }
    normalized
  } 
  // ... rest
}
```

### [nit] useMemo dependency array consistency

The dependency arrays are correctly updated to use `configPaymentMethodType` instead of `paymentMethodType`. This is a good fix for the memoization consistency issue.

**Comment by XyneSpaces** [comment]
## Code Review: PR #1641

### Review Summary

I've reviewed the DynamicFields component changes. This appears to be fixing a payment method type mapping issue.

---

**File:** `src/Components/DynamicFields.res`
**Lines 179-196**

The change introduces `configPaymentMethodType` using `PaymentUtils.getPaymentMethodName` to normalize the payment method type before passing to `SdkConfigParser.getEligibleConnectorsFromPaymentMethods` and `buildSuperpositionBaseContext`.

**Change Pattern:**
```diff
-  let eligibleConnectors = React.useMemo(() => {
-    SdkConfigParser.getEligibleConnectorsFromPaymentMethods(
-      sdkConfigsValue.payment_methods,
-      paymentMethod,
-      paymentMethodType,
-    )
+  let configPaymentMethodType = PaymentUtils.getPaymentMethodName(
+    ~paymentMethodType=paymentMethod,
+    ~paymentMethodName=paymentMethodType,
+  )
+  let eligibleConnectors = React.useMemo(() => {
+    SdkConfigParser.getEligibleConnectorsFromPaymentMethods(
+      sdkConfigsValue.payment_methods,
+      paymentMethod,
+      configPaymentMethodType,
+    )
```

---

**Question/Concern:**

The argument naming in `getPaymentMethodName` call appears potentially swapped:
```rescript
PaymentUtils.getPaymentMethodName(
  ~paymentMethodType=paymentMethod,     // Is this intentional?
  ~paymentMethodName=paymentMethodType, // Or should these be reversed?
)
```

⚠️ **[should-fix]** Please verify that `~paymentMethodType=paymentMethod` and `~paymentMethodName=paymentMethodType` is the intended parameter mapping, as this looks like the arguments might be transposed compared to typical naming conventions.

If this is intentional (perhaps the function signature differs from the parameter names), consider adding a comment explaining the swap to prevent confusion during future maintenance.

---

**Dependency Array Updates:**
✅ Good catch updating the React.useMemo dependency arrays to use `configPaymentMethodType` instead of `paymentMethodType`. This ensures the memoization is correctly invalidated when the normalized value changes.

---

### Overall Assessment
The fix addresses a legitimate issue with payment method type handling. Please confirm the parameter order in the `getPaymentMethodName` call is intentional before merging.

**Comment by XyneSpaces** [comment]
## Review Comments

### DynamicFields - Payment Method Type Normalization

**Line ~179-182 and usage in `src/Components/DynamicFields.res`**:

```rescript
let configPaymentMethodType = PaymentUtils.getPaymentMethodName(
  ~paymentMethodType=paymentMethod,
  ~paymentMethodName=paymentMethodType,
)
```

This fix correctly normalizes `paymentMethodType` using `PaymentUtils.getPaymentMethodName` before passing to downstream functions. This resolves lookup mismatches for bank debit/transfer types that use suffixed names in the SDK config.

**Question**: Does `getPaymentMethodName` handle all edge cases for payment method naming variations? Consider adding unit tests for the normalization function to document the expected mappings:
- `ach_debit` → `ach`
- `multibanco_transfer` → `multibanco`
- etc.

**Dependency array correctness** ✅:
The React hook dependencies were correctly updated from `paymentMethodType` to `configPaymentMethodType` in both `useMemo` calls. This prevents stale closures and ensures proper re-computation when the normalized value changes.

LGTM.

**Comment by XyneSpaces** [comment]
## Review: PR #1641 - fix: paymentMethodType in DynamicFields

### Summary
This PR fixes the payment method type mapping in DynamicFields by introducing `configPaymentMethodType` computed via `PaymentUtils.getPaymentMethodName`, ensuring consistent configuration lookup for eligible connectors and superposition context.

### Findings

✅ **Correct approach**: Using `PaymentUtils.getPaymentMethodName` to normalize the payment method type before passing to downstream functions is the right pattern.

⚠️ **[should-fix] Dependency array completeness**: The dependency array for `superpositionBaseContext` memo includes `configPaymentMethodType` but should also verify if `paymentMethod` itself needs to remain - the original had `paymentMethodType` directly, now using the computed value.

💡 **[nit] Variable naming**: `configPaymentMethodType` could be more descriptive as `normalizedPaymentMethodType` or `paymentMethodTypeKey` to better indicate its purpose as a lookup key.

**Overall**: Clean, focused fix. The normalization pattern improves consistency with how payment methods are resolved elsewhere in the SDK.

**Comment by XyneSpaces** [comment]
## PR Review Summary

### Files Changed: 1 file
- `src/Components/DynamicFields.res`

### Findings

#### [should-fix] Verify getPaymentMethodName behavior
The change introduces `configPaymentMethodType` using `PaymentUtils.getPaymentMethodName` to normalize the payment method type before passing to `SdkConfigParser.getEligibleConnectorsFromPaymentMethods`.

**Consider:** Verify that `getPaymentMethodName` handles all edge cases correctly:
- Empty or null paymentMethodType
- Custom/wallet payment methods that may have different naming conventions
- The mapping should be deterministic and match the SDK config schema

#### [nit] Dependency array consistency
The change updates React.useMemo dependency arrays from `paymentMethodType` to `configPaymentMethodType`. This is correct since `configPaymentMethodType` is derived from both `paymentMethod` and `paymentMethodType`, ensuring the memo recalculates when either changes.

**Overall Assessment:** The fix addresses a potential mismatch between raw payment method types and SDK configuration expectations. The implementation follows existing patterns in the codebase.

---
*Automated review by XyneSpaces*

**Comment by XyneSpaces** [comment]
## Code Review Findings

### [should-fix] Incorrect dependency array in `eligibleConnectors` useMemo

The `eligibleConnectors` memo dependency array uses `configPaymentMethodType` correctly, but verify that `PaymentUtils.getPaymentMethodName` is a pure function (no side effects) since it's called inline during render before the memo.

If `getPaymentMethodName` has any non-deterministic behavior or throws, it could cause unnecessary re-renders or errors. Consider memoizing the `configPaymentMethodType` value itself:

```rescript
let configPaymentMethodType = React.useMemo(() => {
  PaymentUtils.getPaymentMethodName(
    ~paymentMethodType=paymentMethod,
    ~paymentMethodName=paymentMethodType,
  )
}, (paymentMethod, paymentMethodType))
```

This ensures consistent reference equality and prevents redundant calculations.

**Comment by XyneSpaces** [comment]
[nit] Consider adding a defensive log when `configPaymentMethodType` doesn't match any Superposition config entries to aid debugging lookup misses:

```rescript
let eligibleConnectors = React.useMemo(() => {
  let connectors = SdkConfigParser.getEligibleConnectorsFromPaymentMethods(...)
  if connectors->Array.length == 0 {
    Logger.debug("No eligible connectors found for: " ++ configPaymentMethodType)
  }
  connectors
}, ...)
```

This helps identify when normalization produces unexpected values.

**Comment by XyneSpaces** [comment]
[nit] The variable name `configPaymentMethodType` combined with using `PaymentUtils.getPaymentMethodName` creates a naming mismatch that could confuse future maintainers.

`getPaymentMethodName` suggests retrieving a display/localized name (as the API documentation indicates), but here it's being used to normalize a payment method type for configuration lookups. These are semantically different operations — one returns human-readable text, the other returns a canonical key.

**Suggestion:** Consider either:
1. Using a more explicit function like `normalizePaymentMethodType` or `getCanonicalPaymentMethodType` if the function actually performs normalization
2. Renaming the variable to reflect what the value represents (e.g., `canonicalPaymentMethodType` or `normalizedPaymentMethodType`)
3. Adding a clarifying comment explaining why `getPaymentMethodName` is being used for config lookups

**Comment by XyneSpaces** [comment]
## 🔍 Automated Review: DynamicFields Payment Method Type Fix

**PR Scope:** Fixing paymentMethodType normalization for bank debit/transfer types

### ✅ Analysis

**Root Cause Resolution:**
The fix correctly addresses the mismatch between raw payment method types (`ach_debit`, `sepa_transfer`) and the canonical names (`ach`, `sepa`) expected by Superposition and `sdk-configs` API.

**Implementation Quality:**
- Single source of truth via `PaymentUtils.getPaymentMethodName()`
- Consistent application across both `eligibleConnectors` and `superpositionBaseContext`
- Proper dependency updates in React.useMemo hooks

### 💡 Observations

**Hook Dependencies Updated Correctly:**
```rescript
// Before
}, (sdkConfigsValue.payment_methods, paymentMethod, paymentMethodType))

// After  
}, (sdkConfigsValue.payment_methods, paymentMethod, configPaymentMethodType))
```
✅ This prevents stale closures and ensures memoization works correctly.

**Naming Clarity:**
`configPaymentMethodType` clearly indicates this is the normalized variant for config lookups, distinguishing it from the raw `paymentMethodType` prop.

### 🧪 Suggestion

Consider adding a brief comment explaining why normalization is needed:
```rescript
// Normalize payment method type to match Superposition/sdk-configs canonical names
// e.g., "ach_debit" -> "ach", "sepa_transfer" -> "sepa"
```

---
*Automated review by XyneSpaces*

**Comment by XyneSpaces** [comment]
## Automated Code Review Summary

### Overview
This PR fixes `paymentMethodType` usage in `DynamicFields` component by using `PaymentUtils.getPaymentMethodName()` to get the canonical payment method type name from configuration.

### Findings

#### Good Practice ✅
- Fixes a real bug where raw `paymentMethodType` was being passed instead of the canonical config name
- Proper use of `React.useMemo` for derived values
- Dependency arrays updated correctly

#### [should-fix] Variable naming clarity
**Location:** `src/Components/DynamicFields.res` line +179

Consider renaming `configPaymentMethodType` to something more descriptive like `canonicalPaymentMethodType` or `sdkPaymentMethodType` to clarify its purpose vs the raw prop.

```rescript
let canonicalPaymentMethodType = PaymentUtils.getPaymentMethodName(
  ~paymentMethodType=paymentMethod,
  ~paymentMethodName=paymentMethodType,
)
```

This improves readability and helps prevent similar confusion in the future.

**Comment by XyneSpaces** [comment]
[should-fix] The dependency array for `eligibleConnectors` uses `configPaymentMethodType` but `paymentMethodType` is still passed to `getEligibleConnectorsFromPaymentMethods`. This could lead to stale closure issues if `paymentMethodType` changes independently.

```rescript
// Current - inconsistent: dependency uses configPaymentMethodType but function receives raw paymentMethodType
let eligibleConnectors = React.useMemo(() => {
  SdkConfigParser.getEligibleConnectorsFromPaymentMethods(
    sdkConfigsValue.payment_methods,
    paymentMethod,
    configPaymentMethodType,  // Uses normalized value
  )
}, (sdkConfigsValue.payment_methods, paymentMethod, configPaymentMethodType))
```

The change correctly uses `configPaymentMethodType` consistently now - verified against the diff. Consider adding a comment explaining why normalization is needed for maintainability.

**Comment by XyneSpaces** [comment]
[nit] Consider adding debug logging for lookup misses

The fix correctly normalizes `paymentMethodType` for Superposition lookups. Consider adding debug logging when the normalized name doesn't match any config entries to aid future debugging:

```rescript
let eligibleConnectors = React.useMemo(() => {
  let connectors = SdkConfigParser.getEligibleConnectorsFromPaymentMethods(
    sdkConfigsValue.payment_methods,
    paymentMethod,
    configPaymentMethodType,
  )
  
  // Debug helper:
  if connectors->Array.length == 0 && !isProduction {
    Js.Console.warn(
      `No eligible connectors found for ${paymentMethod}/${configPaymentMethodType}`
    )
  }
  
  connectors->Array.map(item => item->JSON.Encode.string)
}, (sdkConfigsValue.payment_methods, paymentMethod, configPaymentMethodType))
```

This would help identify lookup mismatches in development/testing environments.

**Comment by XyneSpaces** [comment]
[should-fix] **Verify configPaymentMethodType mapping for all payment types**

The change introduces `PaymentUtils.getPaymentMethodName` to normalize the `paymentMethodType` value. Verify that this mapping handles all edge cases correctly, particularly:
- Wallet payment methods (Google Pay, Apple Pay)
- BNPL methods (Klarna, Affirm)
- Cryptocurrency methods

The `eligibleConnectors` and `superpositionBaseContext` now depend on this normalized value. Ensure `PaymentUtils.getPaymentMethodName` doesn't return unexpected values for non-card payment methods.

---

**Overall Assessment:** Clean refactor that centralizes payment method type normalization. The useMemo dependency updates correctly reflect the new derived value.

File: src/Components/DynamicFields.res [hunk 1]
```diff
@@ -176,25 +176,29 @@ let make = (
 
   let intentData = paymentMethodListValue.intent_data.intentDataObject
 
+  let configPaymentMethodType = PaymentUtils.getPaymentMethodName(
```
Comment (XyneSpaces) [inline src/Components/DynamicFields.res:179]: Consider wrapping `configPaymentMethodType` in a `useMemo` hook since it's used as a dependency in multiple `useMemo` hooks below (lines 185 and 196). Currently, this value is recreated on every render via `PaymentUtils.getPaymentMethodName()`, which could cause the dependent memos to recalculate unnecessarily even when `paymentMethod` and `paymentMethodType` haven't changed.

Suggested change:
```rescript
let configPaymentMethodType = React.useMemo(() => {
  PaymentUtils.getPaymentMethodName(
    ~paymentMethodType=paymentMethod,
    ~paymentMethodName=paymentMethodType,
  )
}, (paymentMethod, paymentMethodType))
```

**Review by XyneSpaces — COMMENTED** [review]
Performance suggestion regarding the new `configPaymentMethodType` value.

**Comment by XyneSpaces** [comment]
## PR Review Summary

**Review Status:** Reviewed ✓

### Observations

1. **[should-fix]** The `configPaymentMethodType` computation should be wrapped in `React.useMemo` to prevent unnecessary recalculations on every render. Currently, this value is recomputed on each render and used as a dependency in multiple `useMemo` hooks, potentially causing cascading recalculations.

### Positive Notes
- The fix properly normalizes payment method type names by stripping suffixes like `_debit` and `_transfer`
- Good alignment with existing `PaymentUtils.getPaymentMethodName` pattern
- Dependencies in `useMemo` hooks are correctly updated to use the new normalized value

### Suggestions
- Consider adding a unit test to verify the normalization behavior for various payment method type inputs
- Ensure the change is tested with both bank debit and bank transfer payment methods

File: src/Components/DynamicFields.res (hunk 1 above)
Comment (Shivam25092001) [reply src/Components/DynamicFields.res:179]: `getPaymentMethodName(...)` returns a string (primitive), this means referential equality already holds, downstream `useMemo` deps work fine.

**Review by aritro2002 — APPROVED** [review]

**Review by AbhishekChorotiya — APPROVED** [review]
